using_restricted_array.rb

```diff
@@ -7,41 +7,88 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+    i = 0
+    until array[i] == nil
+      puts array[i]
+      i += 1
+    end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  until array[length - 1] < 0
+    return true if array[length - 1] == value_to_find
+    length -= 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  max = array[0]
+  length = length - 1
+  until length == 0
+    if array[length] > max
+      max = array[length]
+    end
+    length -= 1
+  end
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  min = array[0]
+  length = length - 1
+  until length == 0
+    min = array[length - 1] if array[length - 1] < min
+    length -= 1
+  end
+  return min
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until i == length/2
+    array[i] = array[i] + array[length - 1]
+    array[length - 1] = array[i] - array[length - 1]
+    array[i] = array[i] - array[length - 1]
+    i += 1
+    length -= 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  left = 0
+  right = length - 1
+  marker = (right - left)/2
+  until left > right
+    if array[marker] < value_to_find
+      left = marker + 1
+    elsif array[marker] > value_to_find
+      right = marker - 1
+    else
+      return true
+    end
+  end
+  false
```

Comment:
What if there is only one element in the array and it is the value_to_find? In this case, it used to return false. To fix this, add an additional if statement after the loop ends, to check if array[left] is the value_to_find. This was one of the cases I was missing in the algorithm when we reviewed it in class. The slide has since been updated.